### PR TITLE
Refactor Code - Milestone 1

### PR DESCRIPTION
**Pull request by faizi2500**
## In this milestone following tasks are achieved

- [x] YAGNI, KISS, and DRY rules practiced 

- [x] SOLID principles followed

- [x] `Main.rb` file follows single responsibility principle

- [x] Pair programming done for the project

### CODE REVIEW

File: book_module.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+module NewBook
+  def create_book
+    print 'Title:'
+    title = gets.chomp
+    print 'Author:'
+    author = gets.chomp
+    book = Book.new(title: title, author: author)
+    @books << book
+    puts 'Book created successfully'
+  end
```
Comment (mikethreels) [inline book_module.rb]: - [x] good job on cleaning up the `main.rb` file 👍 however this method is not according to the "single responsibility principle" all requests for input and input validation should've been placed in classes/methods whose responsibilities are to request or validate input. I've put a small example below

```
def get_book_info
    print 'Title:'
    title = gets.chomp
    print 'Author:'
    author = gets.chomp
   
    puts 'Book created successfully'
    "and at the end return title and author"
end
def create_book
   book_info = get_book_info
    book = Book.new(title: title, author: author)
    @books << book
end
```

File: person_module.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def create_student
+    input_obj = { 'Age: ' => '', 'Name: ' => '', 'Has Parent Permission: ' => '' }
+    input_obj.each do |key, _value|
+      print key
+      _value = gets.chomp
+    end
+    student = Student.new('Microverse', input_obj[0], input_obj[1])
+    @people << student
+    puts 'Person created successfully'
+  end
```
Comment (mikethreels) [inline person_module.rb]: - [x] good job on cleaning up the `main.rb` file 👍 however this method is not according to the "single responsibility principle" all requests for input and input validation should've been placed in classes/methods whose responsibilities are to request or validate input. Please refer to my previous comment and [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) for example and explanation.

File: person_module.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  def create_teacher
+    input_obj = { 'Age: ' => '', 'Name: ' => '', 'Specialization: ' => '' }
+    input_obj.each do |key, _value|
+      print key
+      _value = gets.chomp
+    end
+    teacher = Teacher.new(input_obj[2], input_obj[0], input_obj[1])
+    @people << teacher
+    puts 'Person created successfully'
+  end
```
Comment (mikethreels) [inline person_module.rb]: - [x] good job on cleaning up the `main.rb` file 👍 however this method is not according to the "single responsibility principle" all requests for input and input validation should've been placed in classes/methods whose responsibilities are to request or validate input. Please refer to my previous comment and [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) for example and explanation.

File: rental_module.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    book_index = gets.chomp.to_i
+    print "Select a person from the folowing list by number (not id)\n"
+    @people.each_with_index { |p, index| puts "#{index}) [#{p.class}] Name: #{p.name}, Age: #{p.age}, ID: #{p.id}" }
+    person_index = gets.chomp.to_i
+    print 'Date: '
+    date = gets.chomp
+    rental = Rental.new(date, @books[book_index], @people[person_index])
+    @rentals << rental
+    puts 'Rental created successfully'
+  end
```
Comment (mikethreels) [inline rental_module.rb]: - [x] good job on cleaning up the `main.rb` file 👍 however this method is not according to the "single responsibility principle" all requests for input and input validation should've been placed in classes/methods whose responsibilities are to request or validate input. Please refer to my previous comment and [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) for example and explanation.

File: rental_module.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  def list_rentals
+    puts 'ID of person:'
+    id = gets.chomp
+    puts 'Rentals:'
+    rentals = @rentals.select { |rental| rental.person.id == id.to_i }
+    rentals.each { |rent| puts "Date: #{rent.date}, Book #{rent.book.title} by #{rent.book.author}" }
+  end
```
Comment (mikethreels) [inline rental_module.rb]: - [x] good job on cleaning up the `main.rb` file 👍 however this method is not according to the "single responsibility principle" all requests for input and input validation should've been placed in classes/methods whose responsibilities are to request or validate input. Please refer to my previous comment and [this article](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/solid.md#best-practices-and-solid) for example and explanation.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Faizii, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book_module.rb (hunk 1 above)
Comment (aliabbani) [reply book_module.rb]: Thank you for your valuable comment, we solved it as you can see in the image below
bookmodule
.

File: person_module.rb (hunk 2 above)
Comment (aliabbani) [reply person_module.rb]: Thank you for your valuable comment, we solved it as you can see in the image below
image

File: person_module.rb (hunk 3 above)
Comment (aliabbani) [reply person_module.rb]: Thank you for your valuable comment, we solved it.

File: rental_module.rb (hunk 4 above)
Comment (aliabbani) [reply rental_module.rb]: Thank you for your valuable comment, we solved it as you can see in the image below
image

File: rental_module.rb (hunk 5 above)
Comment (aliabbani) [reply rental_module.rb]: Thank you for your valuable comment, we solved it as you can see in the image belowimage

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi @aliabbani, 

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**